File: Java/315.count-of-smaller-numbers-after-self.java
/*
 * @lc app=leetcode id=315 lang=java
 *
 * [315] Count of Smaller Numbers After Self
 */

// @lc code=start
/*
    Binary Search Tree, Node{val, count of dup, count of element less than node val}
    先将数组倒序，然后在倒序插入每个元素的过程中，向右边children遍历则加上该node的count + lessCount, 
    代表其比这些元素都要大，可以得到该位置node比其小的元素数目；如果是向左边，则更新lessCount(+1)
    time: O(nlogn) ~ O(n^2)
    time: O(n unique elements)
*/
class Solution {
    class BSTNode {
        int val;
        int count;
        int lessCount;
        BSTNode left;
        BSTNode right;

        public BSTNode(int val) {
            this.val = val;
            count = 1;
        }

        public int smallerSum() {
            return this.count + this.lessCount;
        }
    }

    public List<Integer> countSmaller(int[] nums) {
        List<Integer> res = new ArrayList<>();
        if (nums == null || nums.length == 0) return res;
        res.add(0);
        BSTNode root = new BSTNode(nums[nums.length - 1]);
        for (int i = nums.length - 2; i >= 0; --i) {
            res.add(insert(root, nums[i]));
        }
        Collections.reverse(res);
        return res;
    }

    private int insert(BSTNode root, int num) {
        if (num == root.val) {
            root.count++;
            return root.lessCount;
        } else if (num < root.val) {
            ++root.lessCount;
            if (root.left == null) {
                root.left = new BSTNode(num);
                return 0;
            } else {
                return insert(root.left, num);
            }
        } else {
            if (root.right == null) {
                root.right = new BSTNode(num);
                return root.smallerSum();
            } else {
                return insert(root.right, num) + root.smallerSum();
            }
        }
    }
}
// @lc code=end

File: src/315.count-of-smaller-numbers-after-self.cpp
/*
 * @lc app=leetcode id=315 lang=cpp
 *
 * [315] Count of Smaller Numbers After Self
 */

// @lc code=start
/*
https://paper.dropbox.com/doc/LC315-Count-of-Smaller-Numbers-After-Self--AyHoWxU6VIidxlBz~LZ_txJ6Ag-0ToMbjV1oQmmYKjHm4sFf
对于找一个数组有多少个元素比某个位置的元素小，可以先将数组倒序排列，然后遍历过程中找多少个元素比之前的大。
将数组排序，然后找到对应元素是第几小的rank(rank从1开始)，然后用map记录对应rank元素的出现次数，然后将0 → rank - 1中所有的频率相加(Prefix Sum)即得到比该rank位置小的元素
对于这种prefix sum的操作，可以利用Fenwick Tree来实现，每次遍历一个元素进行Query(rank - 1)的操作，之后在rank位置更新其频率(+1) 
time: O(nlogn)
space: O(unique nums)
*/
class BITree {
public:
    BITree(int n) : prefixSum(n + 1, 0) {}
    int query(int i) {
        int sum = 0;
        while (i > 0) {
            sum += prefixSum[i];
            i -= i & (-i);
        }
        return sum;
    }

    void update(int i, int delta) {
        while (i < prefixSum.size()) {
            prefixSum[i] += delta;
            i += i & (-i);
        }
    }
protected:
    vector<int> prefixSum;
};
class Solution {
public:
    vector<int> countSmaller(vector<int>& nums) {
        vector<int> res;
        set<int> sorted(nums.begin(), nums.end());
        unordered_map<int, int> rankMap;
        int rank = 1;
        for (int num : sorted) {
            rankMap[num] = rank++;
        }
        BITree bitree(rankMap.size());
        for (int i = nums.size() - 1; i >= 0; --i){
            int query = bitree.query(rankMap[nums[i]] - 1);
            res.push_back(query);
            bitree.update(rankMap[nums[i]], 1);
        }
        reverse(res.begin(), res.end());
        return res;
    }
};
// @lc code=end

